zqd/ingest/log.go

```diff
@@ -5,207 +5,138 @@ import (
-			}
-			pipe.Send(&api.LogPostWarning{
-				Type:    "LogPostWarning",
-				Warning: fmt.Sprintf("%s: %s", path, err),
-			})
-			continue
+		if err := p.openFile(zctx, path, cfg); err != nil {
+			return nil, err
 		}
-
-		totalSize += size
-		readCounters = append(readCounters, rc)
-		readers = append(readers, sf)
 	}
+	go p.start(ctx, store)
+	return p, nil
+}
 
-	logBytesRead := func() int64 {
-		var n int64
-		for i := range readCounters {
-			n += readCounters[i].nread
-		}
-		return n
-	}
+func (p *LogOp) openWarning(path string, err error) {
+	p.warnings = append(p.warnings, fmt.Sprintf("%s: %s", path, err))
+}
 
-	zngfile, err := s.CreateFile(filepath.Join(tmpIngestDir, allZngTmpFile))
+func (p *LogOp) openFile(zctx *resolver.Context, path string, cfg detector.OpenConfig) error {
+	info, err := os.Stat(path)
 	if err != nil {
 		return err
 	}
-	zw := zngio.NewWriter(zngfile, zio.WriterFlags{StreamRecordsMax: s.StreamSize()})
-	const program = "sort -r ts | (filter *; head 1; tail 1)"
-	var headW, tailW recWriter
-
-	mux, err := compileLogIngest(ctx, s, readers, program, req.StopErr)
+	if info.IsDir() {
+		return zqe.E(zqe.Invalid, "path is a directory")
+	}
+	f, err := fs.Open(path)
 	if err != nil {
 		return err
 	}
-	d := &logdriver{
-		pipe:         pipe,
-		startTime:    nano.Now(),
-		totalSize:    totalSize,
-		logBytesRead: logBytesRead,
-		writers:      []zbuf.Writer{zw, &headW, &tailW},
-	}
-
-	statsTicker := time.NewTicker(search.StatsInterval)
-	defer statsTicker.Stop()
-	err = driver.Run(mux, d, statsTicker.C)
+	format, err := detector.Format(f)
 	if err != nil {
-		zngfile.Close()
-		os.Remove(zngfile.Name())
-		return err
+		// If an error occur here just log it as a warning and do not add to the
+		// list of readers.
+		p.openWarning(path, err)
+		return f.Close()
 	}
-	if err := zngfile.Close(); err != nil {
-		os.Remove(zngfile.Name())
+	if _, err = f.Seek(0, io.SeekStart); err != nil {
 		return err
 	}
-	if tailW.r != nil {
-		min := nano.Min(tailW.r.Ts, headW.r.Ts)
-		max := nano.Max(tailW.r.Ts, headW.r.Ts)
-		if err = s.SetSpan(nano.NewSpanTs(min, max+1)); err != nil {
-			os.Remove(zngfile.Name())
+	reader := detector.GzipReader(f)
+	reader = io.TeeReader(reader, p)
+	// We can safely ignore here error because error would have been caught in
+	// detector.Format.
+	zr, _ := detector.LookupReader(reader, zctx, format)
+	if zjr, ok := zr.(*ndjsonio.Reader); ok {
+		if err := detector.ConfigureNDJSONReader(zjr, cfg, path); err != nil {
 			return err
 		}
 	}
-	if err := os.Rename(zngfile.Name(), s.DataPath(space.AllZngFile)); err != nil {
-		os.Remove(zngfile.Name())
-		return err
-	}
-	return pipe.Send(&api.LogPostStatus{
-		Type:         "LogPostStatus",
-		LogTotalSize: d.totalSize,
-		LogReadSize:  logBytesRead(),
-	})
+	// Wrap zbuf.Reader in NamedReader (to make file easier to
+	// identify for errors) and WarningReader (so Read warnings propogate to
+	// warning channel).
+	zr = zbuf.NamedReader(zr, path)
```

Comment:
I don't see how NamedReader's String() method could be called.

Reply:
If it's the last reader wrapped then it works. Warning reader should happen before this.

Reply:
I was trying to model the wonder flexibility of io package- the ability to wrap and compose interfaces.

Reply:
Got rid of it.